### PR TITLE
[GPU][DG2] Fix force_implementation build option to work

### PR DESCRIPTION
**Pull request by tuxedcat**
### Details:
 - 

### Tickets:
 - 67491

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp [hunk 1]
```diff
@@ -664,6 +664,13 @@ void reorder_inputs::run(program& p, layout_optimizer& lo, reorder_factory& rf)
 
     auto fmt_map = get_preferred_formats(p, lo);
 
+    for (auto n : p.get_processing_order()) {
+        auto forced_format = lo.get_forced_format(*n);
+        if (forced_format != format::any) {
+            n->set_preferred_output_fmt(0, forced_format);
+        }
+    }
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp:672]: If it is targeting onednn, it is better implemented in `select_preferred_format` pass

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 2]
```diff
@@ -1654,6 +1654,13 @@ impl_types layout_optimizer::get_preferred_impl_type(program_node& node, format
     return preferred_impl;
 }
 
+format layout_optimizer::get_forced_format(program_node& node) {
+    if (!_forcing_map.empty() && _forcing_map.count(node.id()) != 0) {
+        return _forcing_map.at(node.id()).first;
+    }
+    return format::any;
+}
+
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1663]: Please check if it is really necessary from test case.

**Review by isanghao — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp (hunk 1 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp]: Done.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp (hunk 2 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1663]: IMHO, It doesn't have to be a function, but using get_forced_format() function is more readable than using its implementation code directly.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 3]
```diff
@@ -1866,11 +1873,13 @@ void layout_optimizer::select_preferred_formats_for_onednn(program_node& node, d
...
             }
+            auto forced_format = get_forced_format(node);
+            if (forced_format != format::any)
+                dst_fmt = forced_format;
 
-            if (node.get_preferred_output_fmt() == format::any) {
-                for (size_t usr = 0; usr < std::max<size_t>(1, node.get_users().size()); usr++)
+            for (size_t usr = 0; usr < std::max<size_t>(1, node.get_users().size()); usr++)
+                if (node.get_preferred_output_fmt(usr) == format::any)
                     node.set_preferred_output_fmt(usr, dst_fmt);
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1882]: It is related to multi-output port. I don't think it is correct to fix this way.

**Review by isanghao — COMMENTED** [review]
Let's not merge this PR at this moment. Let's merge it when we have actual case for such usage.

**Comment by isanghao** [comment]
This is not fixing any test case at this moment. We may revisit such case when adding test for it.

**Pull request closed without merging**